### PR TITLE
WriteLeaves: Always require the expected revision

### PR DESCRIPTION
**Pull request by pavelkalinnikov**
This change removes the feature of `Set/WriteLeaves` API calls, which allowed to commit leaves to a new "latest" revision rather than a fixed one. This feature is dangerous because it can cause races etc, and it was not used anyway. Additionally, the `revision == 0` meaning was overloaded (depending on the RPC being called): it could mean the "empty tree" revision, or "the latest" revision. Now it has only one meaning.

### Checklist

- [x] I have updated the [CHANGELOG](CHANGELOG.md).
  - Adjust the draft version number according to [semantic versioning](https://semver.org/) rules.
- [x] I have updated [documentation](docs/) accordingly (including the [feature implementation matrix](docs/Feature_Implementation_Matrix.md)).

### CODE REVIEW

File: server/map_rpc_server.go [hunk 1]
```diff
@@ -384,8 +380,8 @@ func (t *TrillianMapServer) getWriteRevision(ctx context.Context, tree *trillian
 	if err != nil {
 		return 0, err
 	}
-	if assertRev != 0 && writeRev != assertRev {
-		return 0, status.Errorf(codes.FailedPrecondition, "can't write to revision %v", assertRev)
+	if writeRev != assertRev {
+		return 0, status.Errorf(codes.FailedPrecondition, "can't write to revision %v: only %d", assertRev, writeRev)
```
Comment (pavelkalinnikov) [inline server/map_rpc_server.go]: Oups, this was added during the debugging phase, I'll restore the original line.

File: CHANGELOG.md [hunk 2]
```diff
@@ -6,9 +6,11 @@ Not yet released; provisionally v1.3.3 (may change).
 
 ### Map Changes
 
-The verifiable map is still experimental. APIs have been deprecated and will be
-deleted in the near future. These changes will not affect the Trillian module
-semantic version due to the experimental status of the Map.
+The verifiable map is still experimental. APIs, such as SetLeaves, have been
+deprecated and will be deleted in the near future. The semantic of WriteLeaves
```
Comment (mhutchinson) [inline CHANGELOG.md]: nit: in English saying "the semantics of X are" is more common than "the semantic of X is".

File: server/map_rpc_server.go (hunk 1 above)
Comment (mhutchinson) [reply server/map_rpc_server.go]: Reminder to please address.

**Comment by gdbelvin** [comment]
Please verify this against Key Transparency locally first

**Comment by pavelkalinnikov** [comment]
KT tests pass locally.

**Pull request merged**